### PR TITLE
Refactor data attributes for form fields

### PR DESCRIPTION
**Pull request by HLeithner**
Reduced duplicated code introduce by #27212 and added missing Fields.

### Summary of Changes
Added a data render function to formfield
Refactor layouts to use preprocessed variable
Added missing fields

### Testing Instructions
Code Review and check if Joomla forms till work especially the new fields. 

### Expected result
Works

### Actual result
Works

### CODE REVIEW

File: layouts/joomla/form/field/contenthistory.php [hunk 1]
```diff
@@ -63,7 +64,11 @@
...
 
 ?>
-<button type="button" onclick="document.getElementById('versionsModal').open()" class="btn btn-secondary" data-toggle="modal">
+<button
+		type="button"
+		onclick="document.getElementById('versionsModal').open()"
+		class="btn btn-secondary"
+		data-toggle="modal"<?php echo $dataAttribute; ?>>
 	<span class="fas fa-code-branch" aria-hidden="true"></span>
 	<?php echo $label; ?>
```
Comment (Quy) [inline layouts/joomla/form/field/contenthistory.php]:
```suggestion
<button
	type="button"
	onclick="document.getElementById('versionsModal').open()"
	class="btn btn-secondary"
	data-toggle="modal"<?php echo $dataAttribute; ?>>
		<span class="fas fa-code-branch" aria-hidden="true"></span>
		<?php echo $label; ?>
```

File: layouts/joomla/form/field/user.php [hunk 2]
```diff
@@ -100,14 +101,6 @@
 {
 	$inputAttributes['placeholder'] = Text::_('JLIB_FORM_SELECT_USER');
 }
-
-if (!empty($dataAttributes))
```
Comment (wilsonge) [inline layouts/joomla/form/field/user.php:104]: Your breaking this one. This was appending to `$inputAttributes`

Comment (HLeithner) [reply layouts/joomla/form/field/user.php:104]: Fixed thanks

File: layouts/joomla/form/field/user.php [hunk 3]
```diff
@@ -117,7 +110,8 @@
 		modal-height="400px"
 		input=".field-user-input"
 		input-name=".field-user-input-name"
-		button-select=".button-select">
+		button-select=".button-select"
+		<?php echo $dataAttribute; ?>>
 	<div class="input-group">
 		<input <?php echo ArrayHelper::toString($inputAttributes); ?> readonly>
```
Comment (SharkyKZ) [inline layouts/joomla/form/field/user.php]: This was on `<input>` element before.

Comment (HLeithner) [reply layouts/joomla/form/field/user.php]: I know but I think it should be on the joomla wc or I'm wrong?

File: layouts/joomla/form/field/media.php [hunk 4]
```diff
@@ -36,20 +36,12 @@
...
 $attr = '';
-$dataAttribute = '';
-
-if (!empty($dataAttributes))
-{
-	foreach ($dataAttributes as $key => $attrValue)
-	{
-		$dataAttribute .= ' ' . $key . '="' . htmlspecialchars($attrValue, ENT_COMPAT, 'UTF-8') . '"';
-	}
-}
```
Comment (wilsonge) [inline layouts/joomla/form/field/media.php:52]: Nothing here is actually using the attr attribute as far as I can see here :) so the data attribute is effectively not being rendered (was a bug before your PR too)

File: layouts/joomla/form/field/user.php (hunk 3 above)
Comment (wilsonge) [reply layouts/joomla/form/field/user.php]: Fancy select is on the input within the WC too. Hard to say which is better - there's an argument for adding support to both to be honest. Sometimes you want fill out the actual field value (i.e. the input), sometimes you want add properties to the WC for custom JS logic

File: layouts/joomla/form/field/media.php (hunk 4 above)
Comment (HLeithner) [reply layouts/joomla/form/field/media.php:52]: It's used in line 157 at the end of the input field

File: layouts/joomla/form/field/user.php (hunk 3 above)
Comment (HLeithner) [reply layouts/joomla/form/field/user.php]: @dgrammatiko any opinion on this?

Comment (dgrammatiko) [reply layouts/joomla/form/field/user.php]: > Sometimes you want fill out the actual field value (i.e. the input), sometimes you want add properties to the WC for custom JS logic

This is very true. The reason you probably want to also fill a nested input in a custom element is to cover forms where the data is populated and the JS is disabled. If there is an input `type=submit` then the form will be submitted but if the input wasn't there then the backend will not get that particular input data which will lead to undesired behaviour. Edge case? Well, I don't know, I thought the core output should be consistent for all possible cases. Even if the core doesn't even use any such case. The Form Fields should be as close to metal as possible they're fundamental building blocks so their support should be solid. At least that was my approach, you can ditch it and dictate things, I mean Joomla is already doing this for the Form fields as they're still hardcoded to Bootstrap and Font Awesome which is definitely wrong if the aim is broad support...

In this case 
```
<input <?php echo ArrayHelper::toString($inputAttributes); ?> <?php echo $dataAttribute; ?> readonly>
```

seems the change that could be merged without any testing

File: layouts/joomla/form/field/media.php (hunk 4 above)
Comment (wilsonge) [reply layouts/joomla/form/field/media.php:52]: I'm blind

File: layouts/joomla/form/field/user.php (hunk 3 above)
Comment (HLeithner) [reply layouts/joomla/form/field/user.php]: Ok I moved it to the input, but if js disabled you can't use it in the input anyway.

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**